Submit code coverage for all builds

mihails-strasuns-sociomantic opened the pull request.

mihails-strasuns-sociomantic committed:
Submit code coverage for all builds

mihails-strasuns-sociomantic commented:
Right I remember now that there was issue with D2 builds because -cov triggers stomping prevention assert

leandro-lucarella-sociomantic commented:
How do you see this issue in the long term? Is there a way to make `-cov` not stomp? Why is the adding of `-cov` making any changes in stomping?

mihails-strasuns-sociomantic commented:
> Is there a way to make -cov not stomp?

I am working on dmd-transitional patch right now

> Why is the adding of -cov making any changes in stomping?

Because it is `-cov` implementation code inside druntime which triggers the assertion (and it is not being used at all unless you use `-cov`)

nemanja-boric-sociomantic commented:
https://github.com/sociomantic-tsunami/ocean/pull/230#issuecomment-326316241
https://github.com/sociomantic-tsunami/ocean/pull/230#issuecomment-326255956
https://github.com/sociomantic-tsunami/ocean/pull/230#issuecomment-326272862

:grumpy:

mihails-strasuns-sociomantic commented:
Yeah, so it is supposed to be in s3

mihails-strasuns-sociomantic committed:
Use latest compiler release for a version

mihails-strasuns-sociomantic commented:
Picked required changes from old @nemanja-boric-sociomantic PR :D

mihails-strasuns-sociomantic commented:
Green

nemanja-boric-sociomantic commented on src/ocean/core/array/Search.d:
```diff
@@ -958,13 +958,13 @@ unittest
    Performs a linear scan of haystack from [0 .. haystack.length$(RP), returning
    a count of the number of elements where pred returns true.
-  
+
```
Not really important, but this should be in a separate commit (or not committed at all).

leandro-lucarella-sociomantic commented on src/ocean/core/array/Search.d:
```diff
@@ -1278,7 +1278,9 @@ unittest
-    return (arr.length >= prefix.length) && (arr[0..prefix.length] == prefix[]);
+    if (arr.length < prefix.length)
+        return false;
+    return arr[0..prefix.length] == prefix[];
```
I would like to see a better explanation for this change in the commit message, at simple sight it looks like rearranging the code. So what's the real problem and what's the intended change should be stated in the commit message.

mihails-strasuns-sociomantic replied:
It is simply re-arranging code indeed. Problem is major code generation bug in compiler (can't find bugzilla number right now). Looks like it doesn't do early return from `&&` expression here when both `-O` and `-cov` are supplied.

leandro-lucarella-sociomantic replied:
I strongly suggest adding a comment here too, otherwise someone could try to rearrange the code again in the future. Ideally a comment with a link to the upstream bug.

mihails-strasuns-sociomantic replied:
Done

mihails-strasuns-sociomantic committed:
Trailing whitespace

mihails-strasuns-sociomantic committed:
Workaround DMD -O -cov code gen bug

Looks like it doesn't do early return if first part of `&&` expression
fails (spec requires to do so
https://dlang.org/spec/expression.html#AndAndExpression) when this
specific compiler flag combination is used.

It was not possible to reproduce same problem with smaller test case.

mihails-strasuns-sociomantic commented:
Updated:
- separate whitespace commit
- more info about DMD bug
